implicit/lmf.pyx

```diff
@@ -239,26 +246,26 @@ def lmf_update(RNGVector rng, floating[:, :] deriv_sum_sq,
                     i = indices[index]
                     for _ in range(n_factors):
                         exp_r += user_vectors[u, _] * item_vectors[i, _]
-                    exp_r = exp(exp_r)
-                    z = (data[index] * exp_r) / (1 + exp_r)
+                    z = sigmoid(exp_r) * data[index]
                     for _ in range(n_factors):
-                        deriv[_] -= z * item_vectors[i, _]
-
+                        deriv[_] =  deriv[_] - z * item_vectors[i, _]
+                
                 # Negative(Sampled) Item Indices exp(y_ui) / (1 + exp(y_ui)) * y_i
                 for _ in range(min(n_items, user_seen_item * neg_prop)):
                     index = rng.generate(thread_id)
                     i = indices[index]
                     exp_r = 0
                     for _ in range(n_factors):
-                        exp_r += user_vectors[u, _] * item_vectors[i, _]
-                    exp_r = exp(exp_r)
-                    z = exp_r / (1 + exp_r)
-                    for _ in range(n_factors):
-                        deriv[_] -= z * item_vectors[i, _]
+                        exp_r = exp_r + (user_vectors[u, _] * item_vectors[i, _])
+                    z = sigmoid(exp_r)
 
+                    for _ in range(n_factors):
+                        deriv[_] = deriv[_] - z * item_vectors[i, _]
                 for _ in range(n_factors):
                     deriv[_] -= reg * user_vectors[u, _]
                     deriv_sum_sq[u, _] += deriv[_] * deriv[_]
-                    user_vectors[u, _] += (lr / sqrt(deriv_sum_sq[u, _])) * deriv[_]
+
+                    # a small constant is added for numerical stability
+                    user_vectors[u, _] += (lr / (sqrt(1e-6 + deriv_sum_sq[u, _]))) * deriv[_]
```

Comment:
I think this will be better to change at line 171-172
```python
        user_vec_deriv_sum = np.zeros((users, self.factors + 2)).astype(np.float32)
        item_vec_deriv_sum = np.zeros((items, self.factors + 2)).astype(np.float32)
```

to
```python
        user_vec_deriv_sum = np.ones((users, self.factors + 2)).astype(np.float32) * EPS
        item_vec_deriv_sum = np.ones((items, self.factors + 2)).astype(np.float32) * EPS
```
where EPS is some value like 1e-3, 1, 1e-6